code/datums/ship.dm

```diff
@@ -97,6 +101,8 @@ GLOBAL_VAR(next_ship_id)
 		C.x_loc = text2num(coords[1])
 		C.y_loc = text2num(coords[2])
 		C.ship = src
+		if(C.attack_data)
+			C.attack_data.ship = src
```

Comment:
move this into the component. This is component specific code in a generic proc. Pass src in the new proc honestly.